File: Tours/src/main/java/com/soa/toursservice/service/ShoppingCartService.java
package com.soa.toursservice.service;

import com.soa.toursservice.model.OrderItem;
import com.soa.toursservice.model.ShoppingCart;
import com.soa.toursservice.model.Tour;
import com.soa.toursservice.model.TourStatus;
import com.soa.toursservice.repository.ShoppingCartRepository;
import com.soa.toursservice.repository.TourRepository;
import org.springframework.stereotype.Service;
import com.soa.toursservice.model.TourPurchaseToken;
import com.soa.toursservice.repository.TourPurchaseTokenRepository;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Service
public class ShoppingCartService {

    private final ShoppingCartRepository shoppingCartRepository;
    private final TourRepository tourRepository;
    private final TourPurchaseTokenRepository tourPurchaseTokenRepository;

    public ShoppingCartService(ShoppingCartRepository shoppingCartRepository,
                               TourRepository tourRepository, TourPurchaseTokenRepository tourPurchaseTokenRepository) {
        this.shoppingCartRepository = shoppingCartRepository;
        this.tourRepository = tourRepository;
        this.tourPurchaseTokenRepository = tourPurchaseTokenRepository;
    }

    public ShoppingCart addTourToCart(String touristUsername, Long tourId) {

        Tour tour = tourRepository.findById(tourId)
                .orElseThrow(() -> new RuntimeException("Tour not found"));

        if (tour.getStatus() != TourStatus.PUBLISHED) {
            throw new RuntimeException("Only published tours can be added to cart");
        }

        ShoppingCart cart = shoppingCartRepository
                .findByTouristUsername(touristUsername)
                .orElseGet(() -> {
                    ShoppingCart newCart = new ShoppingCart();
                    newCart.setTouristUsername(touristUsername);
                    newCart.setTotalPrice(0);
                    return shoppingCartRepository.save(newCart);
                });

        boolean alreadyInCart = cart.getItems()
                .stream()
                .anyMatch(item -> item.getTourId().equals(tourId));

        if (alreadyInCart) {
            throw new RuntimeException("Tour is already in cart");
        }

        OrderItem item = new OrderItem();
        item.setTourId(tour.getId());
        item.setTourName(tour.getName());
        item.setPrice(tour.getPrice());
        item.setShoppingCart(cart);

        cart.getItems().add(item);

        recalculateTotalPrice(cart);

        return shoppingCartRepository.save(cart);
    }

    private void recalculateTotalPrice(ShoppingCart cart) {
        double total = cart.getItems()
                .stream()
                .mapToDouble(OrderItem::getPrice)
                .sum();

        cart.setTotalPrice(total);
    }

    public ShoppingCart getCart(String touristUsername) {

        return shoppingCartRepository
                .findByTouristUsername(touristUsername)
                .orElseGet(() -> {
                    ShoppingCart newCart = new ShoppingCart();
                    newCart.setTouristUsername(touristUsername);
                    newCart.setTotalPrice(0);
                    return shoppingCartRepository.save(newCart);
                });
    }

    public ShoppingCart removeItemFromCart(String touristUsername, Long itemId) {

        ShoppingCart cart = shoppingCartRepository
                .findByTouristUsername(touristUsername)
                .orElseThrow(() -> new RuntimeException("Shopping cart not found"));

        OrderItem itemToRemove = cart.getItems()
                .stream()
                .filter(item -> item.getId().equals(itemId))
                .findFirst()
                .orElseThrow(() -> new RuntimeException("Item not found in cart"));

        cart.getItems().remove(itemToRemove);

        recalculateTotalPrice(cart);

        return shoppingCartRepository.save(cart);
    }

    public List<TourPurchaseToken> checkout(String touristUsername) {

        ShoppingCart cart = shoppingCartRepository
                .findByTouristUsername(touristUsername)
                .orElseThrow(() -> new RuntimeException("Shopping cart not found"));

        if (cart.getItems().isEmpty()) {
            throw new RuntimeException("Shopping cart is empty");
        }

        List<TourPurchaseToken> tokens = new ArrayList<>();

        for (OrderItem item : cart.getItems()) {

            boolean alreadyPurchased = tourPurchaseTokenRepository
                    .existsByTouristUsernameAndTourId(touristUsername, item.getTourId());

            if (!alreadyPurchased) {
                TourPurchaseToken purchaseToken = new TourPurchaseToken();

                purchaseToken.setTouristUsername(touristUsername);
                purchaseToken.setTourId(item.getTourId());
                purchaseToken.setToken(UUID.randomUUID().toString());
                purchaseToken.setPurchasedAt(LocalDateTime.now());

                tokens.add(tourPurchaseTokenRepository.save(purchaseToken));
            }
        }

        cart.getItems().clear();
        cart.setTotalPrice(0);

        shoppingCartRepository.save(cart);

        return tokens;
    }
}

File: Purchase/Purchase.Application/Services/ShoppingCartService.cs
using Purchase.Application.Interfaces;
using Purchase.Domain.Entities;

namespace Purchase.Application.Services;

public class ShoppingCartService
{
    private readonly IShoppingCartRepository _shoppingCartRepository;
    private readonly ITourPurchaseTokenRepository _tourPurchaseTokenRepository;
    private readonly ITourClient _tourClient;

    public ShoppingCartService(
        IShoppingCartRepository shoppingCartRepository,
        ITourPurchaseTokenRepository tourPurchaseTokenRepository,
        ITourClient tourClient)
    {
        _shoppingCartRepository = shoppingCartRepository;
        _tourPurchaseTokenRepository = tourPurchaseTokenRepository;
        _tourClient = tourClient;
    }

    public async Task<ShoppingCart> AddTourToCartAsync(string touristUsername, long tourId)
    {
        var tour = await _tourClient.GetTourPurchaseInfoAsync(tourId);

        if (tour.Status != "PUBLISHED")
        {
            throw new Exception("Only published tours can be added to cart");
        }

        var cart = await _shoppingCartRepository.GetByTouristUsernameAsync(touristUsername);

        if (cart == null)
        {
            cart = new ShoppingCart
            {
                TouristUsername = touristUsername,
                TotalPrice = 0
            };

            cart = await _shoppingCartRepository.AddAsync(cart);
        }

        bool alreadyInCart = cart.Items.Any(item => item.TourId == tourId);

        if (alreadyInCart)
        {
            throw new Exception("Tour is already in cart");
        }

        var item = new OrderItem
        {
            TourId = tour.Id,
            TourName = tour.Name,
            Price = tour.Price,
            ShoppingCart = cart
        };

        cart.Items.Add(item);

        RecalculateTotalPrice(cart);

        return await _shoppingCartRepository.UpdateAsync(cart);
    }

    public async Task<ShoppingCart> GetCartAsync(string touristUsername)
    {
        var cart = await _shoppingCartRepository.GetByTouristUsernameAsync(touristUsername);

        if (cart == null)
        {
            cart = new ShoppingCart
            {
                TouristUsername = touristUsername,
                TotalPrice = 0
            };

            return await _shoppingCartRepository.AddAsync(cart);
        }

        return cart;
    }

    public async Task<ShoppingCart> RemoveItemFromCartAsync(string touristUsername, long itemId)
    {
        var cart = await _shoppingCartRepository.GetByTouristUsernameAsync(touristUsername);

        if (cart == null)
        {
            throw new Exception("Shopping cart not found");
        }

        var itemToRemove = cart.Items
            .FirstOrDefault(item => item.Id == itemId);

        if (itemToRemove == null)
        {
            throw new Exception("Item not found in cart");
        }

        cart.Items.Remove(itemToRemove);

        RecalculateTotalPrice(cart);

        return await _shoppingCartRepository.UpdateAsync(cart);
    }

    public async Task<List<TourPurchaseToken>> CheckoutAsync(string touristUsername)
    {
        var cart = await _shoppingCartRepository.GetByTouristUsernameAsync(touristUsername);

        if (cart == null)
        {
            throw new Exception("Shopping cart not found");
        }

        if (!cart.Items.Any())
        {
            throw new Exception("Shopping cart is empty");
        }

        var tokens = new List<TourPurchaseToken>();

        foreach (var item in cart.Items)
        {
            bool alreadyPurchased = await _tourPurchaseTokenRepository
                .ExistsAsync(touristUsername, item.TourId);

            if (!alreadyPurchased)
            {
                var purchaseToken = new TourPurchaseToken
                {
                    TouristUsername = touristUsername,
                    TourId = item.TourId,
                    Token = Guid.NewGuid().ToString(),
                    PurchasedAt = DateTime.UtcNow
                };

                tokens.Add(await _tourPurchaseTokenRepository.AddAsync(purchaseToken));
            }
        }

        cart.Items.Clear();
        cart.TotalPrice = 0;

        await _shoppingCartRepository.UpdateAsync(cart);

        return tokens;
    }

    public async Task<bool> HasPurchasedAsync(string touristUsername, long tourId)
    {
        return await _tourPurchaseTokenRepository.ExistsAsync(touristUsername, tourId);
    }

    private void RecalculateTotalPrice(ShoppingCart cart)
    {
        cart.TotalPrice = cart.Items.Sum(item => item.Price);
    }
}